### PR TITLE
track when inside tuples during encoding

### PR DESCRIPTION
**Pull request by bluejekyll**
Fixes: #49 

This resurrects and corrects the tests for recursive tuples. It's not a super elegant solution, but I think is correct.

- adds `tuple_depth: usize` to encode and decode
- adds `Option` encode/decode with tuple encoding edge case test

### CODE REVIEW

File: foundationdb/src/tuple/element.rs [hunk 1]
```diff
@@ -256,7 +256,8 @@ impl Decode for String {
 }
 
 impl Encode for Vec<Element> {
-    fn encode<W: Write>(&self, w: &mut W) -> std::io::Result<()> {
+    fn encode<W: Write>(&self, w: &mut W, _in_tuple: bool) -> std::io::Result<()> {
+        // TODO: should this only write the Nested markers in the case of in_tuple?
```
Comment (bluejekyll) [inline foundationdb/src/tuple/element.rs]: I think we only want to write the Nested values in the actual nested case, right? I didn't change this, but it seems like this is wrong, and should follow the same patter in the generated tuple macro.

**Comment by bluejekyll** [comment]
FYI, I just added a change to this PR to drop the single tuple variant

Fixes: #82

File: foundationdb/src/tuple/mod.rs [hunk 2]
```diff
@@ -139,7 +163,6 @@ macro_rules! tuple_impls {
 }
 
 tuple_impls! {
-    1 => (0 T0)
```
Comment (bluejekyll) [inline foundationdb/src/tuple/mod.rs]: By removing this, I think it clarifies single tuple vs. single element in a lot of contexts.

**Comment by bluejekyll** [comment]
Ok, I think this is ready to land.

**Comment by bluejekyll** [comment]
Ok, I've added unary tuple support back, in accordance with #82

**Comment by bluejekyll** [comment]
@yjh0502 and @rushmorem I think after we land this, we're good for the next release.

File: foundationdb/src/tuple/element.rs [hunk 3]
```diff
@@ -459,21 +461,63 @@ impl Decode for i64 {
...
+    T: Decode,
+{
+    fn decode(buf: &[u8], tuple_depth: usize) -> Result<(Self, usize)> {
+        match buf[0] {
+            NIL => {
+                let buf = &buf[1..];
+
+                // custom escape markers are only needed in Nested tuples...
+                if tuple_depth > 1 {
+                    ESCAPE.expect(buf[0])?;
```
Comment (yjh0502) [inline foundationdb/src/tuple/element.rs]: We need a bound check here, too.

File: foundationdb/src/tuple/mod.rs [hunk 4]
```diff
@@ -189,13 +212,13 @@ mod tests {
 
     #[test]
     fn test_malformed_int() {
-        assert!(Tuple::decode(&[21, 0]).is_ok());
-        assert!(Tuple::decode(&[22, 0]).is_err());
-        assert!(Tuple::decode(&[22, 0, 0]).is_ok());
+        assert!(Tuple::decode(&[21, 0], 0).is_ok());
```
Comment (yjh0502) [inline foundationdb/src/tuple/mod.rs]: Could we use `try_from` instead of `decode`, as following tests do?

File: foundationdb/src/tuple/element.rs [hunk 5]
```diff
@@ -459,21 +461,63 @@ impl Decode for i64 {
     }
 }
 
+impl<T> Encode for Option<T>
```
Comment (yjh0502) [inline foundationdb/src/tuple/element.rs]: BTW It looks good!

File: foundationdb/src/tuple/mod.rs (hunk 4 above)
Comment (bluejekyll) [reply foundationdb/src/tuple/mod.rs]: sure, I can make that change.

File: foundationdb/examples/class-scheduling.rs [hunk 6]
```diff
@@ -70,16 +70,16 @@ fn all_classes() -> Vec<String> {
 }
 
 fn init_classes(trx: &Transaction, all_classes: &[String]) {
-    let class_subspace = Subspace::from("class");
+    let class_subspace = Subspace::from(&"class");
```
[the remark this answers is not in the document]
Comment (bluejekyll) [reply foundationdb/examples/class-scheduling.rs]: I needed to add that. `Subspace::from` may not be properly specified, I'll look into it.

File: foundationdb/src/tuple/element.rs (hunk 5 above)
Comment (bluejekyll) [reply foundationdb/src/tuple/element.rs]: thanks!

File: foundationdb/examples/class-scheduling.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (yjh0502) [reply foundationdb/examples/class-scheduling.rs]: Maybe it works only on nightly compiler.. I didn't thought it might not work on stable because my default toolchain is a nightly.. Please ignore this comment, sorry for the noise.

[the remark this answers is not in the document]
Comment (bluejekyll) [reply foundationdb/examples/class-scheduling.rs]: Ok, I just double checked... the CLI cargo build works fine. Apparently it's the VSCode plugin that is reporting an error when it is not. Odd.

**Comment by bluejekyll** [comment]
Thanks for the review. I added the bounds checks... but oddly I couldn't actually figure out a breaking test for them, so there may not have been an issue? Anyway, the new variant is far safer.

**Comment by bluejekyll** [comment]
@rushmorem see https://github.com/bluejekyll/foundationdb-rs/pull/83/commits/83765ffecc72445c9382ea9fa1345b5a526b54f7

This introduces `TupleDepth` with a private constructor and then the `encode_to` and `decode_from` methods. External libraries won't be able to construct `TupleDepth`. I suppose we could even make it more private if we want, so that it can only be constructed in those top level calls. If you look at the bindingstester binary, that has examples of the new interface. Thoughts?

**Comment by bluejekyll** [comment]
@yjh0502 want to give this a final once over?

File: foundationdb/src/tuple/mod.rs [hunk 7]
```diff
@@ -35,6 +36,25 @@ pub enum Error {
     FromUtf8Error(FromUtf8Error),
 }
 
+/// Tracks the depth of a Tuple decoding chain
+#[derive(Copy, Clone)]
+pub struct TupleDepth(usize);
+
+impl TupleDepth {
+    pub(crate) fn new() -> Self {
```
Comment (rushmorem) [inline foundationdb/src/tuple/mod.rs]: This is just brilliant. I love it!

Comment (bluejekyll) [reply foundationdb/src/tuple/mod.rs]: oops, that pub(crate) isn't necessary... let me remove that.

File: foundationdb/src/tuple/mod.rs [hunk 8]
```diff
@@ -35,6 +36,25 @@ pub enum Error {
...
+        TupleDepth(0)
+    }
+
+    fn increment(&self) -> Self {
```
Comment (rushmorem) [inline foundationdb/src/tuple/mod.rs]: Doesn't this need to be public?

File: foundationdb/src/tuple/mod.rs [hunk 9]
```diff
@@ -65,28 +85,40 @@ impl DerefMut for Tuple {
...
     /// Encodes this tuple/elemnt into the associated Write
-    fn encode<W: Write>(&self, _w: &mut W) -> std::io::Result<()>;
+    fn encode<W: Write>(&self, _w: &mut W, tuple_depth: TupleDepth) -> std::io::Result<()>;
+
+    /// Encodes this tuple/elemnt into the associated Write
+    fn encode_to<W: Write>(&self, w: &mut W) -> std::io::Result<()> {
+        self.encode(w, TupleDepth::new())
+    }
+
     /// Encodes this tuple/elemnt into a new Vec
```
Comment (rushmorem) [inline foundationdb/src/tuple/mod.rs]: While you are at it, would you be so kind to fix these typos? Notice `elemnt` in this comment and the other 2 before it.

File: foundationdb/src/tuple/mod.rs [hunk 10]
```diff
@@ -35,6 +36,25 @@ pub enum Error {
...
+/// Tracks the depth of a Tuple decoding chain
+#[derive(Copy, Clone)]
+pub struct TupleDepth(usize);
+
+impl TupleDepth {
+    fn new() -> Self {
+        TupleDepth(0)
+    }
+
+    fn increment(&self) -> Self {
```
Comment (rushmorem) [inline foundationdb/src/tuple/mod.rs]: I don't know what happened to my earlier comment but doesn't `increment` need to be public? Won't someone implementing the traits externally need access to it?

File: foundationdb/src/tuple/mod.rs [hunk 11]
```diff
@@ -224,11 +278,115 @@ mod tests {
...
-            "string".to_vec(),
-            "string".to_string().to_vec()
+            &("one", ("two", 42, ("three", 33))).to_vec(),
+            &[
+                2, 111, 110, 101, 0, 5, 2, 116, 119, 111, 0, 21, 42, 5, 2, 116, 104, 114, 101, 101,
+                0, 21, 33, 0, 0,
+            ]
         );
 
-        assert_eq!("string".to_vec(), ("string",).to_vec());
```
Comment (rushmorem) [inline foundationdb/src/tuple/mod.rs]: Isn't this test still relevant, and even important?

Comment (bluejekyll) [reply foundationdb/src/tuple/mod.rs]: I added a separate set of tests for nested tuples. I can add this back.

File: foundationdb/src/tuple/mod.rs (hunk 10 above)
Comment (bluejekyll) [reply foundationdb/src/tuple/mod.rs]: Yes. You are correct. I will edit.

File: foundationdb/src/tuple/mod.rs (hunk 9 above)
Comment (bluejekyll) [reply foundationdb/src/tuple/mod.rs]: Yep.

File: foundationdb/src/tuple/mod.rs (hunk 11 above)
Comment (rushmorem) [reply foundationdb/src/tuple/mod.rs]: Yes I noticed, but I think this one is unique in that I haven't seen any other test comparing a single element to a tuple with a single element.

Comment (bluejekyll) [reply foundationdb/src/tuple/mod.rs]: technically there is this one :`assert_eq!(&(1,).to_vec(), &1.to_vec());` but it's ok, I've added this back.
